atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java

```diff
@@ -0,0 +1,151 @@
+ */
+public final class LZ4CompressingInputStream extends BufferedDelegateInputStream {
+
+    private static final LZ4Compressor COMPRESSOR = LZ4Factory.fastestInstance().fastCompressor();
+    private static final int DEFAULT_SEED = 0x9747b28c;
+    private static final int DEFAULT_BLOCK_SIZE = 1 << 16; // 64 KB
+
+    private final LZ4BlockOutputStream compressingStream;
+    private final int blockSize;
+    private final byte[] uncompressedBuffer;
+
+    // Position in the compressed buffer while writing
+    private int writeBufferPosition;
+    // Flag to indicate whether this stream has been exhausted.
+    private boolean finished;
+
+    public LZ4CompressingInputStream(InputStream delegate) throws IOException {
+        this(delegate, DEFAULT_BLOCK_SIZE);
+    }
+
+    public LZ4CompressingInputStream(InputStream delegate, int blockSize) throws IOException {
+        super(delegate, Math.max(blockSize, COMPRESSOR.maxCompressedLength(blockSize)));
+        this.blockSize = blockSize;
+        this.uncompressedBuffer = new byte[blockSize];
+        Checksum checksum = XXHashFactory.fastestInstance().newStreamingHash32(DEFAULT_SEED).asChecksum();
+        OutputStream delegateOutputStream = new InternalByteArrayOutputStream();
+        this.compressingStream = new LZ4BlockOutputStream(delegateOutputStream, blockSize, COMPRESSOR, checksum, true);
+        this.finished = false;
+
+        // Flush the LZ4 header
+        compressingStream.flush();
+    }
+
+    @Override
+    protected int refill() throws IOException {
+        if (finished) {
+            return 0;
+        }
+        writeBufferPosition = 0;
+
+        // Read a block of data from the delegate input stream.
+        int bytesRead = ByteStreams.read(delegate, uncompressedBuffer, BUFFER_START, blockSize);
+        if (bytesRead == 0) {
+            // The delegate input stream has been exhausted, so we notify
+            // the compressing stream that there are no remaining bytes.
+            // Lastly, we flush the LZ4 footer into the internal buffer.
+            compressingStream.finish();
+            compressingStream.flush();
+            finished = true;
+        } else {
+            // Write the bytes to the compressing stream and flush it.
+            compressingStream.write(uncompressedBuffer, BUFFER_START, bytesRead);
+            compressingStream.flush();
+        }
+
+        return writeBufferPosition;
+    }
+
+    private void write(int b) throws IOException {
+        ensureCapacity(writeBufferPosition + 1);
+        buffer[writeBufferPosition] = (byte) b;
+        writeBufferPosition++;
+    }
+
+    private void write(byte b[], int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException("Provided byte array b cannot be null.");
+        }
+        if ((off < 0) || (len < 0) || (off + len > b.length)) {
+            throw new IndexOutOfBoundsException();
```

Comment:
is it ever possible to overflow this buffer if the compressed data is bigger than `blockSize`?